src/pairwise.jl

```diff
@@ -0,0 +1,166 @@
+function _pairwise!(::Val{:none}, res::AbstractMatrix, f, x, y, symmetric::Bool)
+    m, n = size(res)
+    for j in 1:n, i in 1:m
+        symmetric && i > j && continue
+
+        # For performance, diagonal is special-cased
+        if f === cor && i == j && x[i] === y[j]
+            # If the type isn't concrete, 1 may not be converted to the right type
+            # and the final matrix will have an abstract eltype
+            # (missings are propagated via the second branch, but NaNs are ignored)
+            res[i, j] = isconcretetype(eltype(res)) ? 1 : one(f(x[i], y[j]))
+        else
+            res[i, j] = f(x[i], y[j])
```

Comment:
Actually in the current implementation `pairwise` calls `collect` on the iterators so they are always `AbstractVector`s. The reason is that currently `eachcol` and `eachrow` are just generators, so `eltype` returns `Any` for them and that requires making a copy of the resulting matrix once we know the final eltype. Allocating vectors of length `m` and `n` is better than making a copy of an `m×n` matrix. But yes I guess I'd better make this code more generic just in case.